Infeasibility anaylsis : make it more changeable

guilpier-code opened the pull request:
Work in progress...

guilpier-code committed:
Renaming infeasibility analyzer : renaming in class Constraint

guilpier-code committed:
Renaming infeasibility analyzer : renaming in class InfeasibleProblemReport

guilpier-code committed:
Simplifying infeasibility analyzer

guilpier-code committed:
New constraint to detect by infeasibility analyzer

flomnes committed:
format

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/simplifying-infeasibility-analyzer

guilpier-code committed:
Infeasibility - to more changeability : simplification + renaming

guilpier-code committed:
Infeasibility - to more changeability : simplifying function StringBetweenAngleBrackets

guilpier-code committed:
Infeasibility - to more changeability : simplifying functions areaName() and timeStep()

guilpier-code committed:
Infeasibility - to more changeability : renaming Constraint::getType() into type()

guilpier-code committed:
Infeasibility - to more changeability : function STSName() ==> simplified and renamed into STSname()

guilpier-code committed:
Infeasibility - to more changeability : remove unnecessary comments

guilpier-code committed:
Infeasibility - to more changeability : simplifying function sortConstraintsByType()

Check on the current constraint's name number of components is useless : it cannot be null.

guilpier-code committed:
Infeasibility - to more changeability : in unit tests, change misleading BC names

guilpier-code committed:
Infeasibility - to more changeability : split constraints logs into 2 parts (suspicious constraints and infeasibility causes)

guilpier-code committed:
Infeasibility - to more changeability : move call to sortConstraintsByType with his friends

guilpier-code committed:
Infeasibility - to more changeability : run clang-format

guilpier-code committed:
Make infeasibily more changeable : move sorting and trimming slack vars from report to analysis

flomnes commented on src/solver/infeasible-problem-analysis/constraint.cpp:
```diff
@@ -36,9 +36,8 @@ const std::string kUnknown = "<unknown>";
-Constraint::Constraint(const std::string& name, const double slackValue):
-    name_(name),
-    slackValue_(slackValue)
+Constraint::Constraint(const std::string& name):
+    name_(name)
```
I think this class becomes useless

guilpier-code replied:
Done

guilpier-code committed:
Make infeasibily more changeable : introducing constraints list to be detected by slack analysis

flomnes commented on src/solver/infeasible-problem-analysis/detected-infeasible-constraints.h:
```diff
@@ -0,0 +1,79 @@
+
+protected:
+    std::string name_;
+    std::string regexId_;
```
Remove protected variable, make corresponding getter regexID virtual

flomnes commented on src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp:
```diff
@@ -31,8 +32,29 @@
+    detectedConstraints_.push_back(std::make_shared<HydroProduction>());
+
+    std::vector<std::string> patterns;
+    std::for_each(detectedConstraints_.begin(), detectedConstraints_.end(),
```
Use accumulate + back_inserter ?

guilpier-code replied:
Now this block is gone, we no longer need **std::for_each** or **std::accumulate**

guilpier-code committed:
Make infeasibily more changeable : use new list of constraint and remove old style

guilpier-code committed:
Make infeasibily more changeable : remove old class Constraint (now useless) + move a header to the right location

guilpier-code committed:
Make infeasibily more changeable : renaming - detected ==> watched (for variables, classes, files)

guilpier-code committed:
Make infeasibily more changeable : split function addSlackVariables into 2 functions

guilpier-code committed:
Make infeasibily more changeable : clang format

flomnes commented on src/solver/infeasible-problem-analysis/report.cpp:
```diff
@@ -21,107 +21,65 @@
+            if (std::regex_search(slackVar->name(), std::regex(cType->regexId())))
+            {
+                constraints_.push_back(cType->clone());
+                constraints_.back()->setConstraintName(slackVar->name());
```
```suggestion
                constraints_.push_back(cType->clone(slackVar->name()));
                continue;
```

guilpier-code replied:
The inner loop was replaced with one line.

guilpier-code committed:
Make infeasibily more changeable : correction due to review

a-zakir commented on src/solver/infeasible-problem-analysis/watched-constraints.cpp:
```diff
@@ -0,0 +1,238 @@
+    return "Weekly BC '" + shortName(splitName_);
+}
+
+std::string WeeklyBC::infeasisibityCause()
```
```suggestion
std::string WeeklyBC::infeasibilityCause()
```

guilpier-code replied:
Done, thanks

a-zakir commented on src/solver/infeasible-problem-analysis/include/antares/solver/infeasible-problem-analysis/watched-constraints.h:
```diff
@@ -0,0 +1,87 @@
+    std::string regexId_;
+    std::vector<std::string> splitName_;
+};
+
```
i think it's not necessary to have all these derived classes!
the only differences are the regex and the infeasibility message

guilpier-code replied:
If you suggest we make one class with 2 member functions, each of them containing a switch or a series of **if** statements, I'm sorry to disagree. 
This is what we used to have, and what I want to avoid : it's contrary to good design rules.
Indeed, it violates **SRP** and fosters big classes/functions.

a-zakir replied:
you only need the regex and the time granularity as ctor inputs

a-zakir replied:
and the later one can be deduced from the constraint name 'after the split

a-zakir replied:
in fact std::string timeStep(std::vector<std::string> splitName) can also return the timstep type (hour, day or weekly)

a-zakir replied:
std::string **WatchedConstraint**::infeasisibity()
{
    return shortName(splitName_) + "' at "+ timeStepType_+ + timeStep(splitName_);
}

a-zakir replied:
and infeasisibityCause seems to be constant for one type ;

guilpier-code replied:
Bodies of **infeasisibity()** and **infeasisibityCause()** can be quite different from a class to another, how would you handle with the messages they return ?

A coded example with 2 constraints would be appreciated.

a-zakir replied:
see my last comment for infeasisibity() method 
and the infeasibilityCause can be (- const std::string member data) obtained from the constructor

guilpier-code replied:
ok, so @a-zakir suggests that we change messages of **infeasibility** and **infeasibility cause**, that is standardize them, in order to have one constraint class instead of one per constraint.
@flomnes : what do you think ?

flomnes replied:
I say you may try what @a-zakir suggests, here's a [possible implementation](https://gist.github.com/flomnes/620f2767094586dbb4180914fb7bd19b)

a-zakir commented on src/solver/infeasible-problem-analysis/watched-constraints.cpp:
```diff
@@ -0,0 +1,238 @@
+    regexId_ = "::weekly::";
+}
+
+std::shared_ptr<WatchedConstraint> WeeklyBC::clone() const
```
this is not a clone, it's a ref to the same object

guilpier-code replied:
**clone()** function were removed, we now use a **factory** instead.

a-zakir commented on src/solver/infeasible-problem-analysis/watched-constraints.cpp:
```diff
@@ -0,0 +1,238 @@
+FictitiousLoad::FictitiousLoad()
+{
+    regexId_ = "^FictiveLoads::";
+}
```
well if you insist, at least this could be in-class initialized

guilpier-code replied:
I do agree

guilpier-code replied:
How would you do that ?

flomnes replied:
Actually, we're not sure `infeasibility` will be called. So there's no need to construct the string in case we need it later.

guilpier-code replied:
Watched constraints no longer have a **regexId_** attribute : it's now associated to the class within in the factory.

a-zakir commented on src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp:
```diff
@@ -53,9 +71,30 @@ void ConstraintSlackAnalysis::run(MPSolver* problem)
+    std::for_each(watchedConstraintTypes_.begin(),
+                  watchedConstraintTypes_.end(),
+                  [&](auto& c) { patterns.push_back(c->regexId()); });
+    std::string constraint_name_pattern{boost::algorithm::join(patterns, "|")};
```
i think regex construction should not be splitted into multiple files!
for the sake of clarity, either here or in "watched constraint" classes

guilpier-code replied:
Due to introduction of a factory, this part was moved, and also simplified. 
See file **watched-constraints.h/.cpp** files.

a-zakir commented on src/solver/infeasible-problem-analysis/include/antares/solver/infeasible-problem-analysis/report.h:
```diff
@@ -38,20 +42,18 @@ class InfeasibleProblemReport
+
+    const std::vector<std::shared_ptr<WatchedConstraint>>& constraintTypes_;
+    const std::vector<const MPVariable*>& slackVariables_;
+    std::vector<std::shared_ptr<WatchedConstraint>> constraints_;
```
3 times std::vector<std::shared_ptr<WatchedConstraint>>
what's went wrong!?

guilpier-code replied:
You right. This part is now simpler, some data members were removed.

a-zakir commented on src/solver/infeasible-problem-analysis/watched-constraints.cpp:
```diff
@@ -0,0 +1,238 @@
+// =======================================
+// Generic constraint logger
+// =======================================
+void WatchedConstraint::setConstraintName(std::string name)
```
```suggestion
void WatchedConstraint::setConstraintName(const std::string& name)
```

a-zakir replied:
or better set the name in the ctor?

guilpier-code replied:
**setConstraintName** was removed, and replaced with a constructor, as you said.

guilpier-code committed:
Make infeasibily more changeable : add use of a factory and remove the clone function

guilpier-code committed:
[skip ci] Make infeasibily more changeable : renaming

guilpier-code committed:
Make infeasibily more changeable : clang format

guilpier-code committed:
Make infeasibily more changeable : avoid useless data member + avoid passing constraint factory

guilpier-code committed:
[skip ci] Make infeasibily more changeable : correct function naming

guilpier-code committed:
Make infeasibility more changeable : simplify constraint factory

guilpier-code committed:
Make infeasibility more changeable : simplifying constraint classes

guilpier-code committed:
Make infeasibility more changeable : fix compilation error on Linux + clang format

guilpier-code committed:
[skip ci] Make infeasibility more changeable ; very small change

guilpier-code committed:
Make infeasibility more changeable : use a std algorithm instead of a for loop

guilpier-code committed:
[skip ci] Make infeasibility more changeable : renaming

guilpier-code committed:
Make infeasibility more changeable : simplify constraint classes yet again

guilpier-code committed:
Make infeasibility more changeable : replace a loop with the use of std::copy_if

guilpier-code committed:
[skip ci] Make infeasibility more changeable : remove useless include

flomnes commented on src/solver/infeasible-problem-analysis/constraint-slack-analysis.cpp:
```diff
@@ -53,9 +59,22 @@ void ConstraintSlackAnalysis::run(MPSolver* problem)
+    ConstraintsFactory factory;
+    std::regex rgx = factory.constraintsFilter();
+    std::copy_if(problem->constraints().begin(),
+                 problem->constraints().end(),
```
```suggestion
    std::ranges::copy_if(problem->constraints(),
```
https://en.cppreference.com/w/cpp/algorithm/ranges/copy

guilpier-code committed:
Make infeasibility more changeable : use of ranges algorithm instead of classic std algorithm

guilpier-code committed:
[skip ci] Make infeasibility more changeable : clang format

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/develop' into fix/infeasibility-more-changeable

flomnes commented on src/solver/infeasible-problem-analysis/include/antares/solver/infeasible-problem-analysis/constraint-slack-analysis.h:
```diff
@@ -52,13 +53,15 @@ class ConstraintSlackAnalysis: public UnfeasibilityAnalysis
-    const std::string constraint_name_pattern = "^AreaHydroLevel::|::hourly::|::daily::|::weekly::|"
-                                                "^FictiveLoads::|^Level::|"
-                                                "^HydroPower::";
+    const unsigned int nbMaxSlackVarsToKeep = 10;
```
```suggestion
    static const unsigned int nbMaxSlackVarsToKeep = 10;
```

flomnes replied:
No need for multiple instances, one is enough

guilpier-code replied:
done

flomnes commented on src/solver/infeasible-problem-analysis/report.cpp:
```diff
@@ -21,107 +21,58 @@
 {
     Antares::logs.error() << "Possible causes of infeasibility:";
-    if (nbConstraintsByType_[ConstraintType::hydro_reservoir_level] > 0)
+    for (const auto& c: uniqueConstraintByType_)
```
You can use `std::unique` to remove duplicates. No need to construct this list, just use `constraints_` and remove duplicates (according to type)

See exemple [here](https://en.cppreference.com/w/cpp/algorithm/unique)

Or you can construct a `std::set` as described above.

guilpier-code replied:
We finally chose to call **std algorithms** in a row instead.

flomnes commented on src/solver/infeasible-problem-analysis/report.cpp:
```diff
@@ -21,107 +21,58 @@
-    for (auto& c: constraints_)
+    std::vector<std::string> pickedTypes;
+    for (const auto& c: constraints_)
     {
-        c.extractComponentsFromName();
-        nbConstraintsByType_[c.type()]++;
+        std::string cType = typeid(*c).name();
+        if (std::find(pickedTypes.begin(), pickedTypes.end(), cType) == pickedTypes.end())
+        {
+            uniqueConstraintByType_.push_back(c);
+            pickedTypes.push_back(cType);
+        }
```
You can use `std::unique`, or just a `std::set` with a custom comparator as described [here](https://gist.github.com/flomnes/f85f7b279ec3b6330eed4176357d1d16). Note that you could also use `std::type_info::hash_code` (type `size_t`) as a key to avoid string comparisons.

flomnes replied:
More simple : return a `std::set<std::string>` containing the possible infeasibility causes. By nature, all elements of a set are unique, so there's no need to remove duplicates.

guilpier-code replied:
We finally chose to call **std algorithms** in a row instead.

flomnes commented on src/solver/infeasible-problem-analysis/watched-constraints.cpp:
```diff
@@ -0,0 +1,217 @@
+
+ConstraintsFactory::ConstraintsFactory()
+{
+    regex_to_ctypes_["::hourly::"] = createHourlyBC;
+    regex_to_ctypes_["::daily::"] = createDailyBC;
+    regex_to_ctypes_["::weekly::"] = createWeeklyBC;
+    regex_to_ctypes_["^FictiveLoads::"] = createFictitiousLoad;
+    regex_to_ctypes_["^AreaHydroLevel::"] = createHydroLevel;
+    regex_to_ctypes_["^Level::"] = createSTS;
+    regex_to_ctypes_["^HydroPower::"] = createHydroProduction;
```
```suggestion
    regex_to_ctypes_["::hourly::"] = std::make_shared<HourlyBC>;
    regex_to_ctypes_["::daily::"] = //;
    regex_to_ctypes_["::weekly::"] = //;
    regex_to_ctypes_["^FictiveLoads::"] = //;
    regex_to_ctypes_["^AreaHydroLevel::"] = //;
    regex_to_ctypes_["^Level::"] = //;
    regex_to_ctypes_["^HydroPower::"] = //;
```

guilpier-code replied:
This isn't what we want, I guess : 
We don't want to associate to a "**regex**" an instance of a watched constraint (for example of **HourlyBC**), but rather a means to construct the type of the watched constraint : for example, to **"::hourly::"**, we'd like to associate the type a function return a fresh instance of **HourlyBC**.

So, I just changed the code to now have : 
```c++
ConstraintsFactory::ConstraintsFactory()
{
    regex_to_ctypes_ = {
      {"::hourly::", [](std::string name) { return std::make_shared<HourlyBC>(name); }},
      {"::daily::", [](std::string name) { return std::make_shared<DailyBC>(name); }},
      ...
      {"^HydroPower::", [](std::string name) { return std::make_shared<HydroProduction>(name); }}};

    auto keyView = std::views::keys(regex_to_ctypes_);
    regex_ids_ = {keyView.begin(), keyView.end()};
}
```
And I removed the creation functions.
Simpler and shorter like that.

flomnes commented on src/solver/infeasible-problem-analysis/include/antares/solver/infeasible-problem-analysis/watched-constraints.h:
```diff
@@ -0,0 +1,104 @@
+class WatchedConstraint
+{
+public:
+    WatchedConstraint(std::string constraintName);
```
```suggestion
    WatchedConstraint(const std::string& constraintName);
```
avoid copies, or use `std::move` later.

guilpier-code replied:
Done

flomnes commented on src/solver/infeasible-problem-analysis/include/antares/solver/infeasible-problem-analysis/watched-constraints.h:
```diff
@@ -0,0 +1,104 @@
+    const std::vector<std::string>& splitName() const;
+
+protected:
+    std::string constraintName_;
```
```suggestion
    std::string fullName_;
```

guilpier-code replied:
Done

flomnes commented on src/solver/infeasible-problem-analysis/watched-constraints.cpp:
```diff
@@ -0,0 +1,217 @@
+    return "* impossible to generate exactly the weekly hydro target";
+}
+
+// --- Constraints factory ---
+std::shared_ptr<WatchedConstraint> createHourlyBC(std::string varName)
+{
+    return std::make_shared<HourlyBC>(varName);
+}
+
+std::shared_ptr<WatchedConstraint> createDailyBC(std::string varName)
+{
+    return std::make_shared<DailyBC>(varName);
+}
+
+std::shared_ptr<WatchedConstraint> createWeeklyBC(std::string varName)
+{
+    return std::make_shared<WeeklyBC>(varName);
+}
+
+std::shared_ptr<WatchedConstraint> createFictitiousLoad(std::string varName)
+{
+    return std::make_shared<FictitiousLoad>(varName);
+}
+
+std::shared_ptr<WatchedConstraint> createHydroLevel(std::string varName)
+{
+    return std::make_shared<HydroLevel>(varName);
+}
+
+std::shared_ptr<WatchedConstraint> createSTS(std::string varName)
+{
+    return std::make_shared<STS>(varName);
+}
+
+std::shared_ptr<WatchedConstraint> createHydroProduction(std::string varName)
+{
+    return std::make_shared<HydroProduction>(varName);
+}
```
```suggestion
```

You don't need all this, `std::make_shared<T>` is already a function.

guilpier-code replied:
Please see [this comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2232#discussion_r1672596805)

flomnes commented on src/solver/infeasible-problem-analysis/watched-constraints.cpp:
```diff
@@ -0,0 +1,217 @@
+
+std::regex ConstraintsFactory::constraintsFilter()
+{
+    return std::regex(boost::algorithm::join(regex_ids_, "|"));
```
```suggestion
    return std::regex(boost::algorithm::join(regex_to_ctypes_ | std::views::keys, "|"));
```
does this work ? If so, `regex_ids_` can be removed.

guilpier-code replied:
Sorry I couldn't make it compile like that.

> does this work ?

yes it does

guilpier-code committed:
Make infeasibility more changeable : simplifying constraint factory

guilpier-code committed:
Make infeasibility more changeable : corrections due to review

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/develop' into fix/infeasibility-more-changeable

flomnes committed:
Fix Sonar issues on #2232 (#2257)

Fix Sonar issues on #2232

flomnes committed:
Minor fixes, use std::unique_ptr, add safety (#2256)

- Use `std::unique_ptr` since we don't share objects
- Remove lambda functions (see
https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique)
- Use `const std::string&`
- Make check on `create`, add safety on no result found
- Remove unused variables in tests

guilpier-code committed:
Make infeasibility more changeable : changing inner class constant into a static const

guilpier-code committed:
Make infeasibility more changeable : trial to simplify method InfeasibleProblemReport :: filterConstraintsToOneByType()

guilpier-code committed:
Make infeasibility more changeable : fix filterConstraintsToOneByType (used to fail on unit tests)

flomnes commented on src/solver/infeasible-problem-analysis/report.cpp:
```diff
@@ -40,26 +39,35 @@ void InfeasibleProblemReport::buildConstraintsFromSlackVars(
+bool sameType(const std::shared_ptr<WatchedConstraint> a, const std::shared_ptr<WatchedConstraint> b)
+{
+    return typeid(*a).name() == typeid(*b).name();
+}
```
Use [`type_index`](https://en.cppreference.com/w/cpp/types/type_index) for these operators, don't re-invent the wheel.

guilpier-code committed:
Make infeasibility more changeable : simplify filterConstraintsToOneByType

flomnes commented on src/solver/infeasible-problem-analysis/report.cpp:
```diff
@@ -40,26 +39,35 @@ void InfeasibleProblemReport::buildConstraintsFromSlackVars(
+    std::stable_sort(constraints_.begin(), constraints_.end(), compTypeName);
+    auto first_duplicate = std::unique(constraints_.begin(), constraints_.end(), sameType);
+    constraints_.erase(first_duplicate, constraints_.end());
+    std::sort(constraints_.begin(), constraints_.end(), compValue);
```
1. Add comments ("eliminate duplicates, etc.")
2. Sort by descending order ? Highest value first

guilpier-code replied:
Done

flomnes reviewed (commented):
`const double` -> `double` should be better (passed by copy)

guilpier-code committed:
Make infeasibility more changeable : fix compilation error

guilpier-code committed:
Make infeasibility more changeable : add comments to make code more understandable

flomnes committed:
Use std::type_index (#2264)

flomnes committed:
Formatting

flomnes committed:
Fix linking

guilpier-code committed:
Make infeasibility more changeable : fix an error in infeasibility causes order in he report
